File: frontend/src/models/Robot.ts
import type { Position } from "@/models/Map";
import Color from "@/models/Color";

export interface PublicRobot {
  id: number;
  name: string;
  position: Position;
  color: Color;
}

export interface Action {
  type: string;
}

export interface Move extends Action {
  direction: string;
}

export interface Scan extends Action {
  distance: number;
}

export interface Wait extends Action {}

export interface Refuel extends Action {}

export interface SolarCharge extends Action {}

export interface FullScan extends Action {}

export interface ActiveRobot {
  id: number;
  name: string;
  color: Color;
  fuel: number;
  maxFuel: number;
  position: Position;
  nextAction: Action | undefined;
  lastResult: string | undefined;
}

export interface ScoreboardEntry {
  name: string;
  color: Color;
  score: number;
}

export function correctTypesFromJson(activeBot: ActiveRobot): ActiveRobot {
  const correctRobot = activeBot;
  correctRobot.color = new Color(correctRobot.color.r, correctRobot.color.g, correctRobot.color.b);

  if (correctRobot.nextAction != null && correctRobot.nextAction.type == "scan") {
    correctRobot.nextAction = correctRobot.nextAction as Scan;
  } else if (correctRobot.nextAction != null && correctRobot.nextAction.type == "move") {
    correctRobot.nextAction = correctRobot.nextAction as Move;
  } else if (correctRobot.nextAction != null && correctRobot.nextAction.type == "wait") {
    correctRobot.nextAction = correctRobot.nextAction as Wait;
  } else if (correctRobot.nextAction != null && correctRobot.nextAction.type == "refuel") {
    correctRobot.nextAction = correctRobot.nextAction as Refuel;
  } else if (correctRobot.nextAction != null && correctRobot.nextAction.type == "solarCharge") {
    correctRobot.nextAction = correctRobot.nextAction as SolarCharge;
  } else if (correctRobot.nextAction != null && correctRobot.nextAction.type == "fullScan") {
    correctRobot.nextAction = correctRobot.nextAction as FullScan;
  } else {
    correctRobot.nextAction = undefined;
  }

  if (correctRobot.lastResult == null) {
    correctRobot.lastResult = undefined;
  }

  return correctRobot;
}

File: backend/src/main/kotlin/xyz/poeschl/roborush/models/Robot.kt
package xyz.poeschl.roborush.models

import com.fasterxml.jackson.annotation.JsonIgnore
import io.swagger.v3.oas.annotations.media.Schema
import net.karneim.pojobuilder.GeneratePojoBuilder
import xyz.poeschl.roborush.configuration.Builder
import xyz.poeschl.roborush.gamelogic.actions.*
import xyz.poeschl.roborush.security.repository.User

@GeneratePojoBuilder(withBuilderInterface = Builder::class)
data class ActiveRobot(
  val id: Long,
  val name: String,
  @JsonIgnore val user: User,
  val color: Color,
  var fuel: Int,
  var position: Position,
  @field:Schema(
    anyOf = [
      MoveAction::class, ScanAction::class, WaitAction::class, RefuelAction::class, SolarChargeAction::class
    ]
  ) var nextAction: RobotAction<*>? = null,
  @field:Schema(
    anyOf = [
      Int::class, ScanAction.ScanResult::class
    ]
  )
  var lastResult: Result? = null,
  @JsonIgnore
  val knownPositions: MutableSet<Position> = mutableSetOf()
) {

  val maxFuel = fuel

  fun canRetrieveFuel() = fuel < maxFuel

  fun addFuel(addedFuel: Int) {
    fuel = (fuel + addedFuel).coerceAtMost(maxFuel)
  }

  fun hasSufficientFuel(expectedFuel: Int) = expectedFuel <= fuel

  fun useFuel(usedFuel: Int) {
    fuel -= usedFuel
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as ActiveRobot

    return id == other.id
  }

  override fun hashCode(): Int = id.hashCode()
}

@GeneratePojoBuilder(withBuilderInterface = Builder::class)
data class PublicRobot(val id: Long, val name: String, val color: Color, var position: Position)

data class ScoreboardEntry(val name: String, val color: Color, val score: Int)

enum class Direction {
  NORTH,
  SOUTH,
  WEST,
  EAST
}
